### PR TITLE
feat: add code collection for all branches

### PR DESCRIPTION
**Pull request by jing-zhi**
#### What this PR does / why we need it:

feat: add code collection for all branches

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?issueFilter__urlQuery=eyJzdGF0ZXMiOls0NDAyLDQ0MDMsNDQwNCw3MTA2LDQ0MDYsNDQwNywyMTAwMSw0NDEyLDQ1MzgsNDQxMywyMzM3Niw0NDE0LDQ0MTUsMjQ3OTYsMjQ3OTcsNDQxNiw0NDE4LDIzMjM2LDQ0MTksNDQyMCw0NDIxLDQ0MjJdLCJhc3NpZ25lZUlEcyI6WyIyMDEyNjIwIl19&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9)

#### Specified Reviewers:

/assign @sfwn @chengjoey 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      feat: add code collection for all branches        |
| 🇨🇳 中文    |      增加全代码采集分析        |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: internal/tools/gittar/metrics/collector.go [hunk 1]
```diff
@@ -165,20 +166,39 @@ func (c *Collector) IterateRepos(req apistructs.GittarListRepoRequest) ([]*Perso
...
+			logrus.Warningf("failed to get all branch for repo: %s, err: %v, use branch: %s as default", repo.Path, err, defaultBranch)
+			branches = append(branches, defaultBranch)
 		}
-		sourceCommit, err := gitRepository.GetBranchCommit(branch)
-		if err != nil {
-			logrus.Errorf("failed to get commit for branch: %s", branch)
-			continue
+		var allSourceCommit []*gitmodule.Commit
+		for _, branch := range branches {
+			if !strutil.HasPrefixes(branch, "master", "Master", "main", "Main", "dev", "Dev", "develop", "Develop", "release", "Release") {
```
Comment (chengjoey) [inline internal/tools/gittar/metrics/collector.go:176]: define `master`, `main` and son on in a variable
like:
var (
  targetBranch = []string{"master","main","dev"}
)

File: internal/tools/gittar/metrics/collector.go [hunk 2]
```diff
@@ -165,20 +166,39 @@ func (c *Collector) IterateRepos(req apistructs.GittarListRepoRequest) ([]*Perso
 		gitRepository.Url = conf.GittarUrl() + "/" + repo.Path
 		gitRepository.IsExternal = repo.IsExternal
 
-		branch, err := gitRepository.GetDefaultBranch()
+		branches, err := gitRepository.GetBranches()
 		if err != nil {
-			logrus.Warningf("failed to get default branch for repo: %s, err: %v, use branch: %s as default", repo.Path, err, defaultBranch)
-			branch = defaultBranch
+			logrus.Warningf("failed to get all branch for repo: %s, err: %v, use branch: %s as default", repo.Path, err, defaultBranch)
+			branches = append(branches, defaultBranch)
```
Comment (chengjoey) [inline internal/tools/gittar/metrics/collector.go:172]: defaultBranch may be duplicated in branches

File: internal/tools/gittar/metrics/collector.go (hunk 1 above)
Comment (chengjoey) [inline internal/tools/gittar/metrics/collector.go:176]: implement a func and add unit tests

File: internal/tools/gittar/metrics/collector.go [hunk 3]
```diff
@@ -165,20 +166,39 @@ func (c *Collector) IterateRepos(req apistructs.GittarListRepoRequest) ([]*Perso
...
-		commits, err := gitRepository.CommitsBetweenDuration(sourceCommit.ID, start, end)
-		if err != nil {
-			logrus.Errorf("failed to list commits for repo: %s, branch: %s", repo.Path, branch)
-			continue
+		commitsSet := make(map[string]bool)
+		var commits []*gitmodule.Commit
+		for _, sourceCommit := range allSourceCommit {
+			curCommits, err := gitRepository.CommitsBetweenDuration(sourceCommit.ID, start, end)
+			if err != nil {
+				logrus.Errorf("failed to list commits for repo: %s, commit-d: %s", repo.Path, sourceCommit.ID)
```
Comment (chengjoey) [inline internal/tools/gittar/metrics/collector.go:191]: commit-id:

File: internal/tools/gittar/metrics/collector.go [hunk 4]
```diff
@@ -165,20 +166,39 @@ func (c *Collector) IterateRepos(req apistructs.GittarListRepoRequest) ([]*Perso
...
+				logrus.Errorf("failed to get curCommit for branch: %s", branch)
+				continue
+			}
+			allSourceCommit = append(allSourceCommit, SourceCommit)
 		}
-		commits, err := gitRepository.CommitsBetweenDuration(sourceCommit.ID, start, end)
-		if err != nil {
-			logrus.Errorf("failed to list commits for repo: %s, branch: %s", repo.Path, branch)
-			continue
+		commitsSet := make(map[string]bool)
```
Comment (chengjoey) [inline internal/tools/gittar/metrics/collector.go:186]: map[string]struct{} is better

File: internal/tools/gittar/metrics/collector.go (hunk 1 above)
Comment (jing-zhi) [reply internal/tools/gittar/metrics/collector.go:176]: ok,yeah,thanks

File: internal/tools/gittar/metrics/collector.go (hunk 4 above)
Comment (jing-zhi) [reply internal/tools/gittar/metrics/collector.go:186]: ok，thanks

File: internal/tools/gittar/metrics/collector.go (hunk 2 above)
Comment (jing-zhi) [reply internal/tools/gittar/metrics/collector.go:172]: ok,thanks

File: internal/tools/gittar/metrics/collector.go (hunk 3 above)
Comment (jing-zhi) [reply internal/tools/gittar/metrics/collector.go:191]: ok,thanks

File: internal/tools/gittar/metrics/collector.go (hunk 1 above)
Comment (jing-zhi) [reply internal/tools/gittar/metrics/collector.go]: ok，thanks

**Review by chengjoey — APPROVED** [review]
PTAL @sfwn

**Pull request merged**